ARMv8-M MPU driver implementation

ioannisg opened the pull request:
This pull request implements the ARMv8-M MPU driver and integrates it to the memory protection system of Zephyr. 
- first set of commits do required refactoring to make it easier to later integrate the V8-M driver. They do not bring in any behavioral changes.
- a separate commit moves the ARMv7-M specific implementation into an internal include file, thus, hiding the code from `arm_mpu.c`.
- the actual ARMv8-M MPU driver implementation is done in a single commit. Similar to ARMv7-M, this is done in an internal include file.
- a last commit adds "UL" in certain integer shift operations.

ioannisg committed:
arch: arm: mpu: move ARMv7m-specific functions in internal header

This commit refactors arm_mpu.c by moving all internal function
implementation that is specific to ARMv7-M architecture to a
separate internal header file. This allows to easily extend
arm_mpu.c to support ARMv8-M by including all ARMv8-M-specific
implementation in a similar internal header.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: type definition for arm mpu attribute container

This commit introduces a type definition for the ARM MPU
region attribute container. This allows to abstract the type
of the attribute container and make the code extendible for
ARMv8-M, where the size and structure of the attribute
container will be different.

Therefore, we can, now, move the definition of the region
data structure in the common arm_mpu.h header.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: refactor _get_region_attr_by_type() function

This commit refactors the _get_region_attr_by_type(.) function,
so it does the following:
 - it takes an arm_mpu_region_attr_t object pointer and use it
   to fill-in the derived region attributes.
 - it performs only the distinguishing between the different MPU
    region types, hiding the actual MPU-specific implementation
	in internal functions.
 - it takes an additional argument (the region base address),
   which allows the function to have a common signature with the
   ARMv8-M version of it (where the attribute derivation needs
   the base address).
 - it returs 0 or succes and a negative non-zero integer
   to indicate an error.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: refactor _get_region_attr_by_conf(.) function

This commit refactors the _get_region_attr_by_conf(.) function,
so it does the following:
 - it is renamed to _get_ram_region_attr_by_conf(), to reflect
   that it is only used to configure RAM MPU regions.
 - it gets an arm_mpu_region_attr_t object structure pointer
   to fill-in with the derived region attributes.
 - it takes an additional argument (the region base address),
   which allows the function to have a common signature with the
   ARMv8-M version of it (where the attribute derivation needs
   the base address).

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
arch: arm: implement ARMv8-M MPU driver

This commit implements and integrates the ARMv8-M MPU driver
into the memory protection system for ARM.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented on include/arch/arm/cortex_m/mpu/arm_mpu_v8m.h:
```diff
@@ -0,0 +1,111 @@
+ */
+
+/* Privileged No Access, Unprivileged No Access */
+//#define NO_ACCESS       0x0
```
These access permissions configuration is not applicable for ARMv8-M. I kept the commented-out code, to make it visible to the reviewing process.

ioannisg commented on include/arch/arm/cortex_m/mpu/arm_mpu_v8m.h:
```diff
@@ -0,0 +1,111 @@
+#define P_RW_U_NA       0x0
+#define P_RW_U_NA_Msk   ((P_RW_U_NA << MPU_RBAR_AP_Pos) & MPU_RBAR_AP_Msk)
+/* Privileged Read Write, Unprivileged Read Only */
+//#define P_RW_U_RO       0x2
```
These access permissions configuration is not applicable for ARMv8-M. I kept the commented-out code, to make it visible to the reviewing process.

ioannisg commented:
The entire kernel Memory Protection tests' suite has been run successfully on nRF52_pca10040 DKs to ensure no bugs are introduced during refactoring.

pizi-nordic commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -35,109 +45,35 @@ static inline u8_t _get_num_regions(void)
  * type (i.e. THREAD_STACK_REGION) and return the correct parameter set.
  */
-static inline u32_t _get_region_attr_by_type(u32_t type, u32_t size)
+static inline arm_mpu_region_attr_t _get_region_attr_by_type(u32_t type,
```
Could we fill-in the provided structure instead of returning one?

pizi-nordic commented on arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h:
```diff
@@ -0,0 +1,161 @@
+ */
+static inline u32_t _size_to_mpu_rasr_size(u32_t size)
+{
+	/* The minimal supported region size is 32 bytes */
```
IMHO we should assert if region is not exactly 32, 64 ... 4G. This way we will avoid risk of setting invalid attributes on memory behind region, which could be considered as security issue. Also this change could be used a verification of our build system, as it should ensure proper location/size of the regions.

ioannisg replied:
The caller of this function is `arm_core_mpu_configure_mem_domain()`, which is invoked by kernel. So, what you ask, is either already done by design, or (if not), we need to enforce the restriction in the kernel and not in the MPU driver.

pizi-nordic replied:
Some MPUs (like ARMv8) does not have such restrictions. I am still opting for assert here as it will check correct usage of the driver.

ioannisg replied:
Now I am a bit confused: this code is, actually, specific to ARMv7...

ioannisg replied:
OK, now I got your point. So, in principle, adding an ASSERT here is not a big deal, so I am going to add it.

ioannisg replied:
This function does not exist for ARMv8-M.

In ARMv8-M 
- such an ASSERT does not make sence, as regions can have any size.
- we, actually, do not need such an ASSERT, because if Kernel miss-configures regions, resulting in overlap, we will get an MPU fault.

pizi-nordic replied:
The fault on ARMv8 will be generated only if the overlapping region will be touched (read, written or executed). In some situations, the fault might be observed very raerly.

ioannisg replied:
>In some situations, the fault might be observed very raerly.

If the fault is not observed, this actually means, we never access at all the overlapping areas, so, this is an "inactive" bug, i.e. it does not get us into any trouble.

Contrary to ARMv7-M, where misconfiguration is "allowed" by the programming model, and we end up configuring regions _not_ in the intended way (so the ASSERT, there make sense).

Now, I agree that we need to make sure that we do not configure overlapping regions, but this challenge shall be addressed "by design". It is very expensive (in run-time) to iterate over all MPU regions and check that they do not overlap with the region we attempt to configure. Instead, we need to implement "adjust region" function and call it in proper places. But this should be done in a subsequent PR.

pizi-nordic replied:
The checking code might be activated using dedicated Kconfig option, which will be turned on during testing only.

ioannisg replied:
Discussed offline: we will need to ensure we never configure overlapping regions, but we won't do this in this PR. I will open an issue for it.

ioannisg replied:
Issue: #8907

pizi-nordic commented on arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h:
```diff
@@ -0,0 +1,161 @@
+	 */
+	(void) base;
+
+	return _get_region_attr(1, ap, 0, 1, 0, 1, 0, size);
```
Could you please rebase this on top of https://github.com/zephyrproject-rtos/zephyr/pull/8786 and check memory attributes?

ioannisg replied:
I guess, I will invite @aurel32 for specifically reviewing the rebase, once #8786 is merged.

aurel32 replied:
@ioannisg, no problem I'll look at the rebase once it's done.

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -14,6 +14,16 @@
+#elif defined(CONFIG_CPU_CORTEX_M23) || \
+	defined(CONFIG_CPU_CORTEX_M33)
+#include <arm_mpu_v8_internal.h>
+#endif
```
Sure, I will add this.

ioannisg replied:
Fixed.

pizi-nordic reviewed (commented):
General: Could we avoid returning structures?

ioannisg commented:
Re-based after the caching fixes are merged; @aurel32, could you take a look to verify that nothing important has been missed?

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v8_internal.h:
```diff
@@ -0,0 +1,117 @@
+
+	MPU->RNR = r_index;
+	r_addr_start = MPU->RBAR & MPU_RBAR_BASE_Msk;
+	r_addr_end = MPU->RLAR & MPU_RLAR_LIMIT_Msk;
```
Good, catch! I am fixing it.

ioannisg replied:
Actually, I am more towards using the modern TT intrinsics for this function. So I am going to re-implement this one.

ioannisg commented on include/arch/arm/cortex_m/mpu/arm_mpu_v8m.h:
```diff
@@ -0,0 +1,111 @@
+
+/* Helper define to calculate the region limit address. */
+#define REGION_LIMIT_ADDR(base, size) \
+	((((base & MPU_RBAR_BASE_Msk) + size) & MPU_RLAR_LIMIT_Msk) \
```
Good catch, here, too; I am fixing it.

ioannisg commented on include/arch/arm/cortex_m/mpu/arm_mpu_v8m.h:
```diff
@@ -0,0 +1,111 @@
+	 | (~(MPU_RLAR_LIMIT_Msk)))
+
+/* Attribute flags for cache-ability */
+#define NORMAL_OUTER_INNER_WRITE_THROUGH_READ_ALLOCATE_TRANS 0x22
```
Yes we can, but we need to define the macros ourselves. ARM CMSIS does not help, here.

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v8_internal.h:
```diff
@@ -0,0 +1,117 @@
+	/* Select the region you want to access */
+	MPU->RNR = index;
+	/* Configure the region */
+	MPU->MAIR0 = region_conf->attr.mair & MPU_MAIR0_Attr0_Msk;
```
By design of the MPU driver we _always_ configure a region with `region_init(.)`, and the MAIR value is passed as argument. So we do not have any need to double-buffer this, IMO.

[As far as I have understood from the manual, once we have configured a region with MAIR, RLAR, RBAR we are allowed to overwrite these registers, assuming, of course, that we have changed the RNR (region number).]

pizi-nordic replied:
>[As far as I have understood from the manual, once we have configured a region with MAIR, RLAR, RBAR we are allowed to overwrite these registers, assuming, of course, that we have changed the RNR (region number).]

I not sure about that. The MAIR registers allow for 8 different attribute sets. If these registers are per-region, then why we can define 8 sets for single region? By my understanding you are only able to select one of the 8 pre-configured attributes in each region.

The https://developer.arm.com/products/architecture/m-profile/docs/100699/latest/initializing-and-configuring-an-mpu seems to confirm that.

ioannisg replied:
Turns out that @pizi-nordic is correct, and the MAIR indirection is performed always when MPU checks are performed, and not only when we configure a region. I will update the implementation here.

ioannisg commented on include/arch/arm/cortex_m/mpu/arm_mpu.h:
```diff
@@ -50,6 +53,16 @@
 #endif /* USERSPACE */
 
+/* Region definition data structure */
+struct arm_mpu_region {
```
The problem with your approach is that the "size" is an attribute in ARMv7, while in ARMv8, it does not even exist (instead, we get the limit address)!

I tried to keep the common parts of the configuration (index, base) outside arm_mpu_region_attr_t, and put all other things inside there. This is back-wards compatible with ARMv7

So, we can eventually have different types of "attributed" depending on the architecture.

I can rename the arguments and provide better documentation to make it easier to understand.

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v8_internal.h:
```diff
@@ -0,0 +1,117 @@
+	attr.rbar = ((1 << MPU_RBAR_XN_Pos) & MPU_RBAR_XN_Msk)
+		| ((ap << MPU_RBAR_AP_Pos) & MPU_RBAR_AP_Msk);
+	attr.mair = NORMAL_OUTER_INNER_WRITE_BACK_WRITE_READ_ALLOCATE_NON_TRANS;
+	attr.rlar = REGION_LIMIT_ADDR(base, size);
```
Sure, I will update the name of the struct argument, to reflect that we always store the limit address, instead of the full attribute list that goes to RLAR.

Update to `r_limit`.

ioannisg replied:
Also, I update the `mair` attribute to `mair_idx` to reflect that this, now, holds only an index value for the MAIR indirection. Updated the documentation in the struct definition, too.

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v8_internal.h:
```diff
@@ -0,0 +1,117 @@
+{
+	arm_mpu_region_attr_t attr;
+
+	attr.rbar = ((1 << MPU_RBAR_XN_Pos) & MPU_RBAR_XN_Msk)
```
Well, here, I think it can stay, because, in the definition of the struct we have:
`struct arm_mpu_region_attr {
	/* Attributes belonging to RBAR */
	u8_t rbar: 5;`

So the documentation of the field indicates clearly that this is just the attributes that go to RBAR; not the base address.

And we always use this as: `attr.rbar`, in the source code.

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v8_internal.h:
```diff
@@ -0,0 +1,145 @@
+ */
+static void _region_init(u32_t index, struct arm_mpu_region *region_conf)
+{
+	ARM_MPU_SetRegion(
```
Here, I used, directly the ARM CMSIS provided function to set up the MPU region.

ioannisg committed:
arch: arm: adapt region_init(.) to use arm_mpu_region_attr structure

This commit adapts the _region_init(.) function so it can use
the rasr field of the arm_mpu_region_attr structure, when it
configures the RASR register.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
>General: Could we avoid returning structures?

@pizi-nordic Refactored once more, so the relevant functions do not return structures, but, instead, fill in structure containers.

pizi-nordic commented on arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h:
```diff
@@ -0,0 +1,225 @@
+	r_addr_start = MPU->RBAR & MPU_RBAR_ADDR_Msk;
+	r_size_lshift = ((MPU->RASR & MPU_RASR_SIZE_Msk) >>
+			MPU_RASR_SIZE_Pos) + 1;
+	r_addr_end = r_addr_start + (1 << r_size_lshift) - 1;
```
1ul << ...

BTW: You can just compare base address with `(base + size - 1) & ~((1ul << shift) - 1)`.

ioannisg replied:
OK about 1UL.
I admit, the suggestion confuses me more :)

pizi-nordic commented on arch/arm/core/cortex_m/mpu/arm_mpu_v8_internal.h:
```diff
@@ -0,0 +1,139 @@
+	u32_t ap_attr, u32_t base, u32_t size)
+{
+	p_attr->rbar = ap_attr  & (MPU_RBAR_XN_Msk | MPU_RBAR_AP_Msk);
+	p_attr->mair_idx = MPU_MAIR_INDEX_SRAM;
```
Should we unconditionally overwrite mair_idx here?

ioannisg replied:
Good point.
The function is used to re-configure RAM regions so this is safe to do.
I can document more clearly that _get_region_attr_by_conf() is used for RAM regions, if you suggest

pizi-nordic commented on arch/arm/core/cortex_m/mpu/arm_mpu_v8_internal.h:
```diff
@@ -0,0 +1,139 @@
+	u32_t ap_attr, u32_t base, u32_t size)
+{
+	p_attr->rbar = ap_attr  & (MPU_RBAR_XN_Msk | MPU_RBAR_AP_Msk);
+	p_attr->mair_idx = MPU_MAIR_INDEX_SRAM;
```
Should we unconditionally overwrite mair_idx here?

ioannisg replied:
Yeap, same point as above.

ioannisg committed:
arch: arm: mpu: explicitly add UL in numerical shift operations

This commit adds a UL suffix after numerical digits to ensure
the correctness of certain shift operations.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@pizi-nordic could you take a final look into this one?

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -267,7 +160,10 @@ void arm_core_mpu_configure(u8_t type, u32_t base, u32_t size)
-	region_conf.attr = _get_region_attr_by_type(type, size);
+
+	if (_get_region_attr_by_type(&region_conf.attr, type, base, size)) {
+		return;
```
Sure, I can add this function into an ASSERT block.

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu.c:
```diff
@@ -489,8 +339,8 @@ static int arm_mpu_init(struct device *arg)
 	size = (u32_t)&__app_ram_end - (u32_t)&__app_ram_start;
-	region_conf.attr =
-		_get_region_attr_by_type(THREAD_APP_DATA_REGION, size);
+	_get_region_attr_by_type(&region_conf.attr, THREAD_APP_DATA_REGION,
```
Same comment as above; I will add the ASSERT.

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h:
```diff
@@ -0,0 +1,225 @@
+	MPU->RBAR = (region_conf->base & MPU_RBAR_ADDR_Msk)
+				| MPU_RBAR_VALID_Msk | index;
+	MPU->RASR = region_conf->attr.rasr | MPU_RASR_ENABLE_Msk;
+	SYS_LOG_DBG("[%d] 0x%08x 0x%08x",
```
Sure, this will  be part of a subsequent PR. I will track this on github.

ioannisg replied:
Tracked under #8018

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v7_internal.h:
```diff
@@ -0,0 +1,225 @@
+		return REGION_4G;
+	}
+
+	size = 1UL << (32 - __builtin_clz(size - 1));
```
Good point; I will investigate this.

ioannisg replied:
Update comment: I submitted #9398 which tries to optimize _size_to_mpu_rasr_size(), so it only uses a single call to __builtin_clz()

ioannisg commented on arch/arm/core/cortex_m/mpu/arm_mpu_v8_internal.h:
```diff
@@ -0,0 +1,139 @@
+	 * SRAM region(s): Attribute-1
+	 */
+	MPU->MAIR0 =
+		((MPU_MAIR_ATTR_FLASH & MPU_MAIR0_Attr0_Msk) >>
```
Good catch, my bad. The correct code should be:

	MPU->MAIR0 =
		((MPU_MAIR_ATTR_FLASH << MPU_MAIR0_Attr0_Pos) &
			MPU_MAIR0_Attr0_Msk)
		|
		((MPU_MAIR_ATTR_SRAM << MPU_MAIR0_Attr1_Pos) &
			MPU_MAIR0_Attr1_Msk)	;

ioannisg committed:
arch: arm: add ASSERT in _get_region_attr_by_type

This commit adds an __ASSERT line in _get_region_attr_by_type,
to signal an error if the supplied MPU region type is not valid.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@agross-linaro , @galak could you also take a look at the PR?

nashif commented:
@AdithyaBaglody please review
